src/gluonnlp/model/bert.py

```diff
@@ -30,13 +30,212 @@
+      - **context_vec** : Symbol or NDArray
+        Shape (query_length, batch_size, context_vec_dim)
+      - **att_weights** : Symbol or NDArray
+        Attention weights of multiple heads.
+        Shape (batch_size, num_heads, query_length, memory_length)
+    """
+    def __init__(self, units, num_heads, dropout=0.0, use_bias=True,
+                 weight_initializer=None, bias_initializer='zeros',
+                 prefix=None, params=None):
+        super().__init__(prefix=prefix, params=params)
+        self._num_heads = num_heads
+        self._use_bias = use_bias
+        self._dropout = dropout
+        self.units = units
+        with self.name_scope():
+            if self._use_bias:
+                self.query_bias = self.params.get('query_bias', shape=(self.units,),
+                                                  init=bias_initializer)
+                self.key_bias = self.params.get('key_bias', shape=(self.units,),
+                                                init=bias_initializer)
+                self.value_bias = self.params.get('value_bias', shape=(self.units,),
+                                                  init=bias_initializer)
+            weight_shape = (self.units, self.units)
+            self.query_weight = self.params.get('query_weight', shape=weight_shape,
+                                                init=weight_initializer,
+                                                allow_deferred_init=True)
+            self.key_weight = self.params.get('key_weight', shape=weight_shape,
+                                              init=weight_initializer,
+                                              allow_deferred_init=True)
+            self.value_weight = self.params.get('value_weight', shape=weight_shape,
+                                                init=weight_initializer,
+                                                allow_deferred_init=True)
+            self.dropout_layer = nn.Dropout(self._dropout)
+
+    def _collect_params_with_prefix(self, prefix=''):
+        # the registered parameter names in v0.8 are the following:
+        # prefix_proj_query.weight, prefix_proj_query.bias
+        # prefix_proj_value.weight, prefix_proj_value.bias
+        # prefix_proj_key.weight, prefix_proj_key.bias
+        # this is a temporary fix to keep backward compatibility, due to an issue in MXNet:
+        # https://github.com/apache/incubator-mxnet/issues/17220
+        if prefix:
+            prefix += '.'
+        ret = {prefix + 'proj_' + k.replace('_', '.') : v for k, v in self._reg_params.items()}
+        for name, child in self._children.items():
+            ret.update(child._collect_params_with_prefix(prefix + name))
+        return ret
+
+    # pylint: disable=arguments-differ
+    def hybrid_forward(self, F, qkv, valid_len, query_bias, key_bias, value_bias,
+                       query_weight, key_weight, value_weight):
+        in_bias = F.concat(query_bias, key_bias, value_bias, dim=0)
```

Comment:
Since u concat everything, why don't we just create a large bias and weight?

Reply:
that leads to different optimizer.step

Reply:
that's fair. we can try in the future and see if it does not affect the convergence.